bayes_kit/drghmc.py

```diff
@@ -0,0 +1,471 @@
+    (model parameters) and rho represents momentum.
+
+    To produce a new draw, partially refresh the momentum and generate a proposed draw
+    with Hamiltonian dynamics. If accepted, the proposed draw and its log density are
+    returned; if rejected, a new proposal is generated. This process is repeated until
+    a maximum number of proposals is reached or when probabilistic delayed rejection
+    returns early.
+
+    This sampler requires non-increasing leapfrog stepsizes such that subsequent
+    proposals are generated with (equal or) more accurate Hamiltonian dynamics. This
+    allows for excellent sampling from multiscale distributions: large stepsizes are
+    used in wide, flat regions while smaller stepsizes are used in narrow,
+    high-curvature regions.
+
+    Implementation based on Modi C, Barnett A, Carpenter B. Delayed rejection
+    Hamiltonian Monte Carlo for sampling multiscale distributions. Bayesian Analysis.
+    2023. https://doi.org/10.1214/23-BA1360.
+    """
+
+    def __init__(
+        self,
+        model: GradModel,
+        proposals: int,
+        leapfrog_stepsizes: list[float],
+        leapfrog_stepcounts: list[int],
+        damping: float,
+        metric_diag: Optional[VectorType] = None,
+        init: Optional[VectorType] = None,
+        seed: Optional[Seed] = None,
+        prob_retry: bool = True,
+    ):
+        """Initialize the DrGhmcDiag sampler.
+
+        Args:
+            model: model with log density and gradient
+            proposals: maximum number of proposal attempts
+            leapfrog_stepsizes: list of non-increasing leapfrog stepsizes
+            leapfrog_stepcounts: list of number of leapfrog steps
+            damping: Generalized HMC momentum damping factor in (0, 1]
+            metric_diag: diagonal of the metric. Defaults to identity metric.
+            init: initialize Markov chain. Defaults to draw from standard normal.
+            seed: seed for numpy rng. Defaults to non-reproducible rng.
+            prob_retry: retry another proposal probabilistically. Defaults to True.
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        self._proposals = self._validate_propoals(proposals)
+        self._leapfrog_stepsizes = self._validate_leapfrog_stepsizes(leapfrog_stepsizes)
+        self._leapfrog_stepcounts = self._validate_leapfrog_stepcounts(
+            leapfrog_stepcounts
+        )
+        self._damping = self._validate_damping(damping)
+        self._metric = metric_diag or np.ones(self._dim)
+        self._rng = np.random.default_rng(seed)
+        self._theta = (
+            init
+            if (init is not None and init.shape != (0,))
+            else self._rng.normal(size=self._dim)
+        )
+        self._rho = self._rng.normal(size=self._dim)
+        self._prob_retry = prob_retry
+
+        # When generating a draw, the cache is used in the joint_logp() and leapfrog()
```

Comment:
I am indeed using this list like a stack for my cache.

Additionally, it is important to note that the cache is used in two ways. Firstly, it is used when generating a single draw as we attempt multiple proposals and recursively compute ghost points. Second, it is used across draws, depending on whether or not the proposed draw is accepted or rejected.

I will condense the inline code comment, but think this point of nuance is important. What are your thoughts?

Reply:
Because of the recursive computation of ghost points, it is not trivial to set up a cache. This is why I originally included such a long inline comment.

Reply:
I would suggest saying what you say above in the comment.  

I don't think you need to say "if rejected, the cache retains the log density and gradient of the current draw."  You can just say that it caches the log density and gradient of the latest draw (that will be last draw or current proposal depending on acceptance).